cogeo_mosaic/backend/base.py

```diff
@@ -0,0 +1,70 @@
+"""cogeo_mosaic.backend.base: base Backend class."""
+
+import abc
+from contextlib import AbstractContextManager
+from typing import Dict, Optional
+
+from cogeo_mosaic import version as mosaic_version
+from cogeo_mosaic.backend.utils import get_hash
+from cogeo_mosaic.utils import create_mosaic
+
+
+class BaseBackend(AbstractContextManager):
+    mosaic_def: Optional[Dict]
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs):
+        """Connect to backend"""
+
+    def __enter__(self):
+        """Support using with Context Managers"""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers"""
+
+    def metadata(self) -> Dict:
+        """Retrieve Mosaic metadata
+
+        Returns
+        -------
+        MosaicJSON as dict without `tiles` key.
+        """
+        if self.mosaic_def is None:
+            return {}
+
+        return {k: v for k, v in self.mosaic_def.items() if k != "tiles"}
+
+    @abc.abstractmethod
+    def tile(self, x: int, y: int, z: int):
```

Comment:
I think using input param of bbox is more flexible (all XYZ tiles have defined bbox but not all bbox have a specific XYZ tile).  This gives more flexibility in terms of caching strategies -- if you know you want to do a lot of tile requests within a specific area you can cache the asset lookup for that entire area.

Reply:
1. Can we be certain that there won't be float precision issues where we ask for one xyz tile and actually retrieve 9? 

2). I feel that it really depends on the internals. If the internals are all in terms of mercator tiles, then it's going to be simplest to just pass down the xyz index. If we pass down a `bbox` that doesn't align with a mercator tile, what does it do? Find all the mercator tiles that overlap with the bbox? Return a custom image with the given pixel density for that `bbox`?

Reply:
> I feel that it really depends on the internals. If the internals are all in terms of mercator tiles, then it's going to be simplest to just pass down the xyz index. If we pass down a bbox that doesn't align with a mercator tile, what does it do? Find all the mercator tiles that overlap with the bbox? Return a custom image with the given pixel density for that bbox?

☝️ totally agree with this. 

I think we could add an other method to retrieve asset from `bbox, z`

Reply:
I added this to a "future todo" list in the original comment, and we can take it up in a future PR.